FEXCore/Source/Interface/Core/LookupCache.h

```diff
@@ -16,6 +16,88 @@
 
 namespace FEXCore {
 
+struct GuestToHostMap {
+  std::recursive_mutex WriteLock;
+
+  struct LockToken {
+    std::lock_guard<std::recursive_mutex> Lock;
+  };
+
+  [[nodiscard]]
+  LockToken AcquireLock() {
+    return LockToken {std::lock_guard {WriteLock}};
+  }
+
+  struct BlockLinkTag {
+    uint64_t GuestDestination;
+    FEXCore::Context::ExitFunctionLinkData* HostLink;
+
+    bool operator<(const BlockLinkTag& other) const {
+      if (GuestDestination < other.GuestDestination) {
+        return true;
+      } else if (GuestDestination == other.GuestDestination) {
+        return HostLink < other.HostLink;
+      } else {
+        return false;
+      }
+    }
+  };
+
+  // Use a monotonic buffer resource to allocate both the std::pmr::map and its members.
+  // This allows us to quickly clear the block link map by clearing the monotonic allocator.
+  // If we had allocated the block link map without the MBR, then clearing the map would require slowly
+  // walking each block member and destructing objects.
+  //
+  // This makes `BlockLinks` look like a raw pointer that could memory leak, but since it is backed by the MBR, it won't.
+  std::pmr::monotonic_buffer_resource BlockLinks_mbr;
+  using BlockLinksMapType = std::pmr::map<BlockLinkTag, FEXCore::Context::BlockDelinkerFunc>;
+  fextl::unique_ptr<std::pmr::polymorphic_allocator<std::byte>> BlockLinks_pma;
+  BlockLinksMapType* BlockLinks;
+
+  // TODO: Should be shared across threads... Also BlockLinks, perhaps?
+  fextl::robin_map<uint64_t, uint64_t> BlockList;
+
+  GuestToHostMap();
+  ~GuestToHostMap();
+
+  // Adds to Guest -> Host code mapping
+  void AddBlockMapping(uint64_t Address, void* HostCode, const LockToken&) {
+    [[maybe_unused]] auto Inserted = BlockList.emplace(Address, (uintptr_t)HostCode).second;
+    // NOTE: If this was inserted twice, we've probably raced against another thread to compile this block. Just ignore this one
+    // TODO: Should reset CodeBuffer cursor in that case...
```

Comment:
Not sure how ordering will work with this pr and multiple entrypoint merges, but this will need to be done for the latter

Reply:
What sort of ordering is needed? This is still protected by the global lock, so there should be no conflict I think. (Taking the lock is a bit subtle - it needs to happen midway through CompileCode but the acquired `unique_lock` is returned to the caller!)

Reply:
Apologies I meant merge ordering :)

Reply:
The latest version of this PR also changes the insert-if-new to an entry override (though it only happens in exceptional cases here), so this works out now.